modules/tensor_mechanics/src/scalarkernels/PeriodicStrain.C

```diff
@@ -0,0 +1,135 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PeriodicStrain.h"
+
+// MOOSE includes
+#include "Assembly.h"
+#include "PeriodicStrainUserObject.h"
+#include "MooseVariableScalar.h"
+#include "SystemBase.h"
+#include "RankTwoTensor.h"
+#include "RankFourTensor.h"
+
+registerMooseObject("TensorMechanicsApp", PeriodicStrain);
+
+template <>
+InputParameters
+validParams<PeriodicStrain>()
+{
+  InputParameters params = validParams<ScalarKernel>();
+  params.addClassDescription("Periodic Strain Scalar Kernel");
+  params.addRequiredParam<UserObjectName>("periodic_strain",
+                                          "The name of the PeriodicStrainUserObject");
+
+  return params;
+}
+
+PeriodicStrain::PeriodicStrain(const InputParameters & parameters)
+  : ScalarKernel(parameters),
+    _pst(getUserObject<PeriodicStrainUserObject>("periodic_strain")),
+    _pst_residual(_pst.getResidual()),
+    _pst_jacobian(_pst.getJacobian()),
+    _components(_var.order())
+{
+  if (_var.order() != FIRST && _var.order() != THIRD && _var.order() != SIXTH &&
+      _var.order() != NINTH)
+    mooseError("PerdiodicStrain ScalarKernel is only compatible with FIRST, THIRD, SIXTH,"
+               "and NINTH order scalar variables. Please change order of the scalar"
+               "variable acoording to the problem dimension.");
+
+  assignComponentIndices(_var.order(), _mesh.dimension());
+}
+
+void
+PeriodicStrain::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+  for (_i = 0; _i < re.size(); ++_i)
+    re(_i) += _pst_residual(_components[_i].first, _components[_i].second);
+}
+
+void
+PeriodicStrain::computeJacobian()
+{
+  DenseMatrix<Number> & ke = _assembly.jacobianBlock(_var.number(), _var.number());
+  for (_i = 0; _i < ke.m(); ++_i)
+    for (_j = 0; _j < ke.m(); ++_j)
+      ke(_i, _j) += _pst_jacobian(_components[_i].first,
+                                  _components[_i].second,
+                                  _components[_j].first,
+                                  _components[_j].second);
+}
+
+void
+PeriodicStrain::computeOffDiagJacobian(unsigned int /*jvar*/)
+{
+}
+
+void
+PeriodicStrain::assignComponentIndices(Order var_order, unsigned int dim)
+{
+  if (dim == 1)
+    if (var_order == FIRST)
```

Comment:
check if dimension and order match up in the constructor and use only order here

Reply:
and make it a switch with a `default:` block that throws an error